DeviceSourceManager

PolygonalSun opened the pull request:
This PR has the initial code for the DeviceSourceManager, a manager that keeps track of specific devices and provides the user with easier to understand context when getting input.

deviceEnum.ts:
- Added enums for device types and inputs for known gamepads (xbox, ps4, and switch)

deviceInputSystem.ts:
- Increased _MAX_KEYCODES to 255 to account for international keyboard layouts
- Added MOUSE_DEVICE and mouse specific ternary operator to account for chance that a mouse's pointerId is not 1 (spec no longer guarantees this).
- Fixed an issue where touch pointers were no longer being registered after initial one
- Removed a throw from pollInput to allow for new change to input data access
- Added observeInput callback to enable use of event-driven updates for Keyboard, Mouse, and Touch

deviceSourceManager.ts:
- Users can now get input from the input system via the getInput function.  This function is not pollInput as I felt that getInput was less simpler terminology.
- There are 4 different observables (connect-before/after, disconnect-before/after) to allow for better customization of the connection process.  While they have no major use in the DSM, they will provide higher levels with some event control.
- All gamepad detection methods are based off of the device names given by the Gamepad API.  Considerations will need to be made for other platforms.
- With the getInput command, you can call it with the device type and input enum and get the first connected input value or you can specify a slot (only applicable to gamepads) and get a specific devices value.

To use this code, the minimum code that you need is to create an instance of the DeviceSourceManager and have a getInput call somewhere in your game/render loop.

Example:
```
var dsm = new BABYLON.DeviceSourceManager(scene.getEngine());
...
scene.onBeforeRenderObservable.add(() => {
     if (dsm.getInput(BABYLON.DeviceType.Xbox, BABYLON.XboxInputs.A) == 1) {
          // Do something while A is pressed
     }
});
```

PolygonalSun commented on src/DeviceInput/deviceInputSystem.ts:
```diff
@@ -15,6 +15,8 @@ export class DeviceInputSystem implements IDisposable {
     /** KEYBOARD_DEVICE */
     public static readonly KEYBOARD_DEVICE: string = "Keyboard";
+    /** MOUSE_DEVICE */
+    public static readonly MOUSE_DEVICE: string = "Mouse";
```
I can do that.

PolygonalSun commented on src/DeviceInput/deviceInputSystem.ts:
```diff
@@ -15,6 +15,8 @@ export class DeviceInputSystem implements IDisposable {
+    public static readonly MOUSE_DEVICE: string = "Mouse";
 
     /**
      * Observable to be triggered when a device is connected
```
I was talking with @bghgary about this yesterday and the plan is to revert these back to callbacks.

PolygonalSun commented:
> My understanding is that this API is entirely polling based, specifically to get current input state. It seems like for many situations this may be hard to use. For example:
> 
> 1. When a key is pressed, I want to add an object to a scene - do I need to manually track key state and look for a change across two frames and then react to that?
> 2. As a pointer is moved across the screen, I want to rotate an object - do I need to track pointer position deltas across frames and apply that delta to a rotation?
> 
> I would have though we'd want to include a more event driven model, or at the very least for polling have a way of querying deltas across frames without the consumer having to manage this themselves. Am I missing anything, or are these scenarios not well covered with the current proposal?
> 
> Edit: It's possible I'm not understanding how this is supposed to work. In your example above, you check whether the Xbox input A's state is `1` and your comment says "Do something when A is **hit**". Would that example be doing something only when A transitions from a non-pressed to pressed state, or would it do something every frame as long as A is in a pressed state? I guess if it is the latter, then how would you expect to do the former (perform an action when A transitions from non-pressed to pressed)?

This is a polling based system.  For #1, I'm not 100% sure that I understand the scenario.  With the code as is, you should be able to just listen for input using onBeforeRenderObservable and add your object there.  This may not be what you're asking so I'd like to confirm things with you offline. For #2, the current system just keeps track of absolute position and click status.  For this scenario, you'd need to calculate and track deltas manually.  If you want, I could add additional support for things like movement deltas without major issue.

Drigax commented on src/DeviceInput/InputDevices/deviceEnums.ts:
```diff
@@ -0,0 +1,189 @@
+ */
+export enum PointerInputs
+{
+    /** Horizontal Axis */
```
What drives this specific pointer input enum?

PolygonalSun replied:
This enum is used with any Touch or Mouse input.  Touch inputs use the first three values while Mouse inputs use all of them.

ryantrem commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,218 @@
+     * @param inputIndex Index of device input
+     * @returns Current value of input
+     */
+    public getInput(type: DeviceType, inputIndex: number, deviceSlot: number = this._firstDevice[type]): Nullable<number> {
```
Correct usage of this function seems not very discoverable or enforceable (e.g. how do you know you are supposed to pass one of the input enum values in for `inputIndex`? You can pass any old number in for `inputIndex` and the code will compile without errors). I think this could be improved with conditional types:
```typescript
type DeviceInputs<T extends DeviceType> =
    T extends DeviceType.Keyboard | DeviceType.GenericDevice ? number :
    T extends DeviceType.Mouse | DeviceType.Touch ? PointerInputs :
    T extends DeviceType.DualShock ? DualShockInputs :
    T extends DeviceType.Xbox ? XboxInputs :
    T extends DeviceType.Switch ? SwitchInputs :
    never;

public getInput<T extends DeviceType>(deviceType: T, inputIndex: DeviceInputs<T>, deviceSlot: number = this._firstDevice[type]): Nullable<number> {
    ...
}
```

This way, when the first parameter (`deviceType`) is known at compile time, then the type of `inputIndex` is both more discoverable and also enforced at compile time. Here is what you see in intellisense when you start typing the function:
![image](https://user-images.githubusercontent.com/5084643/79697392-648d1080-8237-11ea-8c0c-74a96d7cf216.png)
Since `deviceType` has not been specified, `inputIndex` is unconstrained, and intellisense shows exactly the same thing as what it does with your existing implementation. Here is what you see when you've specified `DeviceType.Mouse` for the `deviceType` parameter:
![image](https://user-images.githubusercontent.com/5084643/79697423-a5852500-8237-11ea-9a6f-51bd1094e119.png)
Notice that the `inputIndex` parameter has now been constrained to be of type `PointerInputs`, which is much more discoverable to someone trying to use this API. Further, once `deviceType` has been specified as `DeviceType.Mouse`, then providing anything *other* than `PointerInputs` for the `inputIndex` parameter results in a compile error:
![image](https://user-images.githubusercontent.com/5084643/79697520-26442100-8238-11ea-9551-e5e652794e6f.png)

PolygonalSun replied:
I've implemented the DeviceInputs type to enforce these conditional types

ryantrem commented:
> This is a polling based system. For #1, I'm not 100% sure that I understand the scenario. With the code as is, you should be able to just listen for input using onBeforeRenderObservable and add your object there. This may not be what you're asking so I'd like to confirm things with you offline. For #2, the current system just keeps track of absolute position and click status. For this scenario, you'd need to calculate and track deltas manually. If you want, I could add additional support for things like movement deltas without major issue.
> 
> For the XboxInputs.A example, the comment may be a bit misleading. It should perform whatever action is in the if statement block, each frame, as long as the A button remains pressed. For the alternative, the user would have to add additional code to their query to handle single button presses (eg. add a bool to check that a button was pressed and the execute an action if the button set the bool to true and is currently 0).

For my first example above, I just meant that when the `A` keyboard key (for example) goes from up state to down state, I want to add an object to the scene. I think ultimately all my examples are examples of wanting frame-to-frame deltas. For things like an axis (e.g. pointer position), the delta is really obvious (it moved x pixels). I think we can apply the same thing to buttons though (the delta would be 0 if the button has the same state, either down or up, between frames; the delta would be 1 if it went from not pressed to pressed; the delta would be -1 if it went from pressed to not pressed). I think to make the programming model easy, we should do one of two things (or both):
1. In addition to being able to query the current input state, also be able to query the input state delta since the last frame (for any device type + input index), *and* be able to get a list of input indexes (per device type) that have changed since the last frame (this would make it easy to see that on a device with many input indices, like a keyboard, that specifically the `A` key has changed since the last frame).
2. Expose events for when an input index value changes.

cc @bghgary

ryantrem commented on src/DeviceInput/InputDevices/deviceEnums.ts:
```diff
@@ -0,0 +1,189 @@
+ * Enum for All Pointers (Touch/Mouse)
+ */
+export enum PointerInputs
+{
```
Nit: Move opening curly brace to same line as enum declaration.

PolygonalSun replied:
Fixed

PolygonalSun commented:
@ryantrem @bghgary I believe all of the requested changes should be included now.

PolygonalSun commented:
Here's a sample script that should demonstrate some of the new features:
```
var createScene = function () {

    // This creates a basic Babylon Scene object (non-mesh)
    var scene = new BABYLON.Scene(engine);

    // This creates and positions a free camera (non-mesh)
    var camera = new BABYLON.FreeCamera("camera1", new BABYLON.Vector3(0, 5, -10), scene);
    // Set to false to disable events (false is the default)
    var test = new BABYLON.DeviceSourceManager(scene.getEngine(), true);

    test.onBeforeDeviceConnectedObservable.add((device) => {
        const type = device["deviceType"];
        const slot = device["deviceSlot"];
        switch(type)
        {
            case BABYLON.DeviceType.Keyboard:
                console.log("Keyboard connected.");
                break;
            case BABYLON.DeviceType.Mouse:
                console.log("Mouse connected.");
                break;
            case BABYLON.DeviceType.Touch:
                console.log("Touch detected with id of " + slot);
                break;
            case BABYLON.DeviceType.Xbox:
                console.log("Xbox controller connected in Slot " + slot);
                break;
            case BABYLON.DeviceType.DualShock:
                console.log("Dualshock controller connected in Slot " + slot);
                break;
            case BABYLON.DeviceType.Switch:
                console.log("Switch controller connected in Slot " + slot);
                break;
            case BABYLON.DeviceType.Generic:
            default:
                console.log("Unknown device connected in Slot " + slot);
                break;
        }
    });

    // Test for getDeviceSource
    test.onAfterDeviceConnectedObservable.add((device) => {
        if (device["deviceType"] == BABYLON.DeviceType.Keyboard) {
            test.getDeviceSource(BABYLON.DeviceType.Keyboard).onInputChangedObservable.add((source) => {
                console.log("Input KeyCode: " + source["inputIndex"]);
            });
        }
    });

    scene.onBeforeRenderObservable.add(() => {
        if (test.getInput(BABYLON.DeviceType.Xbox, BABYLON.XboxInput.A, 0) == 1)
        {
            console.log("Player 1: A");
        }
        else if (test.getInput(BABYLON.DeviceType.Xbox, BABYLON.XboxInput.A) == 1)
        {
            console.log("First available: A");
        }
    });
    return scene;
};
```

RaananW commented:
Would it be possible to add an observable when input actually changed and have the user listen to this, instead of running it on the beforeRender loop?
If I understand @ryantrem correctly, this is what he expected the API to look like (in the first comment) - more event driven. 
I know it is much simplified, but the WebXR motion controller API is also polling based. I added a few ways for the user to monitor changes, which you might be able to use here. And again, I understand this is a much broader API. 

And just for me to understand - I noticed you removed observables and switched to callback-based API. Is there a specific reason for that?

bghgary commented on src/DeviceInput/deviceInputSystem.ts:
```diff
@@ -71,24 +92,22 @@ export class DeviceInputSystem implements IDisposable {
 
         if (!device) {
-            throw `Unable to find device ${deviceName}`;
+            return null;
```
It's not clear to me why this needs to change.

PolygonalSun replied:
Originally, this was changed to account for checking for input before a device exists but there are checks outside of this so I put the throw back as it just never throw under *normal* circumstances.

bghgary commented on src/DeviceInput/deviceInputSystem.ts:
```diff
@@ -169,43 +194,56 @@ export class DeviceInputSystem implements IDisposable {
                 pointer[evt.button + 2] = 0;
             }
-            if (evt.pointerId != 1) // Don't unregister the mouse
+            if (evt.pointerType != "mouse") // Don't unregister the mouse
```
nit: Comments that say exactly what the code does is meanless IMO. Maybe explain why you don't unregister the mouse.

PolygonalSun replied:
Changed comment to explain why we don't want to remove the mouse.

bghgary commented on src/DeviceInput/InputDevices/deviceEnums.ts:
```diff
@@ -0,0 +1,199 @@
+/**
+ * Type to handle enforcement of inputs
+ */
+export type DeviceInputs<T extends DeviceType> =
```
nit: this isn't an enum, but it is in a file called deviceEnums

PolygonalSun replied:
I can pull this out into its own file called deviceTypes.

bghgary commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,308 @@
+     * @param deviceSlot "Slot" or index that device is referenced in
+     * @returns Current value of input
+     */
+    public getInput<T extends DeviceType>(type: T, inputIndex: DeviceInputs<T>, deviceSlot: number = this._firstDevice[type]): Nullable<number> {
```
Is this function still useful? It is duplicate functionality with getDeviceSource and then calling getInput?

ryantrem replied:
Yea I find this rather confusing. I would vote to remove it to simplify the API.

PolygonalSun replied:
I originally wanted to leave this to keep with the initial idea of simply asking the DSM for an input but using the getDeviceSource function is functionally the same thing.  Removed.

PolygonalSun replied:
So I was testing the removal of this and it breaks my original ideal of single line poll-and-forget.  Before, you could just keep asking and either get nothing or get input.  Now, it will fail if the device you're listening to hasn't been added yet.  This can be worked around by just checking for the device before polling.  I do want to revisit this but for now I can leave it out.

bghgary replied:
I'm not sure I understand. You can use the other function to do it all in one line no?

PolygonalSun replied:
You can, if a DeviceSource exists for it.  Otherwise, you're trying to get input from a null object.

bghgary replied:
With newer versions of TypeScript, it is easy to do this in one-line:
```javascript
deviceSourceManager.getDeviceSource(...)?.getInput(...)
```

PolygonalSun replied:
That works for TypeScript but it looks like it isn't available yet for JavaScript (proposal for it is finished though so it could be coming soon.).

ryantrem commented on src/DeviceInput/InputDevices/deviceEnums.ts:
```diff
@@ -0,0 +1,199 @@
+    /** Switch Controller */
+    Switch = 5,
+    /** Generic */
+    Generic = 6
```
If we are going to keep `Generic`, can we set its value to `Number.MAX_VALUE` or something so that when we add additional device types in the future, `Generic` can remain at the very end of the enum?

PolygonalSun replied:
With the way that we've implemented keeping track of the devices, I feel like setting this to max wouldn't be the best course of action.  We use the number of enums in DeviceType to create and allocate an array and then copy the data into their corresponding areas using the associated enum's assigned number.

bghgary replied:
Maybe make Generic = 0 and shift everything else?

PolygonalSun replied:
I'd be fine with that.

ryantrem commented on src/DeviceInput/InputDevices/deviceEnums.ts:
```diff
@@ -0,0 +1,199 @@
+/**
+ * Type to handle enforcement of inputs
+ */
+export type DeviceInputs<T extends DeviceType> =
```
If all the enums are not plural, and this type resolves to one of those enums, then it seems like it should also not be plural. That is: `DeviceInput<T extends DeviceType>`

PolygonalSun replied:
I'll change this to DeviceInput

ryantrem commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,308 @@
+    /**
+     * Observable to handle device input changes per device
+     */
+    public onInputChangedObservable = new Observable<{ inputIndex: number, previousState: Nullable<number>, currentState: Nullable<number> }>();
```
```suggestion
    public onInputChangedObservable = new Observable<{ inputIndex: DeviceInput<T>, previousState: Nullable<number>, currentState: Nullable<number> }>();
```

PolygonalSun replied:
Changed

ryantrem commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,308 @@
+     * Observable to handle device input changes per device
+     */
+    public onInputChangedObservable = new Observable<{ inputIndex: number, previousState: Nullable<number>, currentState: Nullable<number> }>();
+    private _deviceInputSystem: DeviceInputSystem;
```
`readonly`?

PolygonalSun replied:
Added

ryantrem commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,308 @@
+        /** Name of device to be used by DeviceInputSystem */
+        public deviceName: string,
+        /** Type of device */
+        public deviceType: DeviceInputs<T>,
```
The type of `deviceType` is `DeviceType`, not `DeviceInputs<T>`. `DeviceInputs<T>` resolves to one of the input enums that corresponds to the `DeviceType`.

PolygonalSun replied:
Changed

ryantrem commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,308 @@
+     * @param inputIndex index of specific input on device
+     * @returns Input value from DeviceInputSystem
+     */
+    public getInput(inputIndex: T): Nullable<number> {
```
`inputIndex` should be of type `DeviceInputs<T>` since it is the enum corresponding to the `DeviceType`.
```suggestion
    public getInput(inputIndex: DeviceInputs<T>): Nullable<number> {
```

PolygonalSun replied:
Changed

ryantrem commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,308 @@
+        return this._deviceInputSystem.pollInput(this.deviceName, inputIndex);
+    }
+
+    /**
+     * Add specific point to array of touch points (DeviceType.Touch only)
+     * @param name Name of Specific Touch Point
+     */
+    public addTouchPoints(name: string) {
+        if (this.deviceType == DeviceType.Touch) {
+            this._touchPoints.push(name);
+        }
+    }
+
+    /**
+     * Remove specific point from array of touch points (DeviceType.Touch only)
+     * @param name Name of Specific Touch Point
+     */
+    public removeTouchPoints(name: string) {
+        if (this.deviceType == DeviceType.Touch) {
+            const touchIndex = this._touchPoints.indexOf(name);
+            this._touchPoints.splice(touchIndex, 1);
+        }
+    }
```
I think these are only publicly exposed so that `DeviceSourceManager` can set them. This is a bit different from what I had imagined. I thought all the state would still be managed by `DeviceSourceManager`, and `DeviceSource` would just provide a typesafe way to interact with a specific device managed by `DeviceSourceManager`. I think doing that would make a bit more sense, and also probably making `DeviceSource` a nested class of `DeviceSourceManager` so that `DeviceSource` can call private methods of `DeviceSourceManager`. This way the public contract will be very clear and not provide access to private state.

PolygonalSun replied:
My initial idea was to just keep track of how many touch inputs there were and keep them in a manageable set of slots so that the user doesn't need to keep up with a bunch of changing pointer ids.  For the time being, I've ripped out all of that code and will be testing things without it.

ryantrem commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,308 @@
+     * @param deviceSlot "Slot" or index that device is referenced in
+     * @returns DeviceSource object
+     */
+    public getDeviceSource(deviceType: DeviceType, deviceSlot: number = this._firstDevice[deviceType]): Nullable<DeviceSource<DeviceType>> {
```
This will not actually get us the type safety we are aiming for. We need a generic type parameter for the conditional type to work:
```suggestion
    public getDeviceSource<T extends DeviceType>(deviceType: T, deviceSlot: number = this._firstDevice[deviceType]): Nullable<DeviceSource<T>> {
```

PolygonalSun replied:
Changed

ryantrem commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,308 @@
+     * @param deviceType Enum specifiying device type
+     * @returns Array of DeviceSource objects
+     */
+    public getDeviceSources(deviceType: DeviceType): ReadonlyArray<DeviceSource<DeviceType>> {
```
```suggestion
    public getDeviceSources<T extends DeviceType>(deviceType: T): ReadonlyArray<DeviceSource<T>> {
```

PolygonalSun replied:
Added.

ryantrem commented on src/DeviceInput/deviceInputSystem.ts:
```diff
@@ -1,30 +1,51 @@
+ * Pointer device string
+ * @hidden
+ */
+const POINTER_DEVICE: string = "Pointer";
```
Any reason you define these on one line, and export them on a different line? I can't see any reason not to just do:
```suggestion
export const POINTER_DEVICE: string = "Pointer";
```
And remove the export statements on lines 21-26.

PolygonalSun replied:
I originally did this because of a validation issue in the build process but never reverted them after just using ```@hidden```.  Changed.

PolygonalSun replied:
These strings have actually been removed.

ryantrem commented on src/DeviceInput/index.ts:
```diff
@@ -1 +1,3 @@
 export * from "./deviceInputSystem";
```
Do we need have some imports here before the exports to ensure these new types don't get stripped out? cc @bghgary

bghgary replied:
Probably. If types don't get used, they won't make it into the js without an additional import.

PolygonalSun replied:
Added import

ryantrem commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,308 @@
+                const deviceType = this._getDeviceTypeFromName(deviceName);
+                const deviceSlot = this._getDeviceSlot(deviceName);
+
+                if (deviceType == DeviceType.Keyboard || deviceType == DeviceType.Mouse || deviceType == DeviceType.Touch) {
```
Why is this restricted to a specific set of device types? Should't any of these have the option of raising events?

PolygonalSun replied:
I had originally added these checks because we don't have event based support for gamepads.  In the code, we don't call the onInputChanged function for gamepads so removing this if statement won't affect gamepads.

PolygonalSun commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,308 @@
+    /**
+     * Observable to handle device input changes per device
+     */
+    public onInputChangedObservable = new Observable<{ inputIndex: number, previousState: Nullable<number>, currentState: Nullable<number> }>();
```
Added

PolygonalSun commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,308 @@
+     * @param deviceSlot "Slot" or index that device is referenced in
+     */
+    constructor(deviceInputSystem: DeviceInputSystem,
+        /** Name of device to be used by DeviceInputSystem */
+        public deviceName: string,
+        /** Type of device */
+        public deviceType: DeviceInputs<T>,
+        /** "Slot" or index that device is referenced in */
+        public deviceSlot: number = 0) {
```
Added readonly

PolygonalSun commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,308 @@
+            case DeviceType.Xbox:
+            case DeviceType.Switch:
+            case DeviceType.Generic:
+                let i = 0;
+                let first = -1;
+
+                while (first < 0 && i < this._devices[type].length) {
+                    if (this._devices[type][i]) {
+                        first = i;
+                    }
+                    i++;
+                }
+
+                this._firstDevice[type] = first;
```
Changed.

PolygonalSun commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,308 @@
+        else if (deviceName == MOUSE_DEVICE) {
+            return DeviceType.Mouse;
+        }
+        else if (deviceName.search(POINTER_DEVICE) !== -1) {
```
Changed.

PolygonalSun commented on src/DeviceInput/deviceInputSystem.ts:
```diff
@@ -1,30 +1,51 @@
+     * Callback to be triggered when event driven input is updated
      */
-    public onDeviceDisconnectedObservable = new Observable<string>();
+    public observeInput: (deviceName: string, inputIndex: number, previousState: Nullable<number>, currentState: Nullable<number>) => void;
```
Changed.

PolygonalSun commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,308 @@
+     * @returns Slot Number
+     */
+    private _getDeviceSlot(deviceName: string): number {
+        const splitName = deviceName.split("-");
```
I've converted everything to just use type and slot so string concatenation shouldn't be an issue anymore.

PolygonalSun commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,308 @@
+     * @param deviceName Name of Device from DeviceInputSystem
+     * @returns DeviceType enum value
+     */
+    private _getDeviceTypeFromName(deviceName: string): DeviceType {
```
Moved to DeviceInputSystem and renamed to _getGamepadDeviceType.  This is because I removed the unnecessary if statements for Keyboard and Pointers.  The only thing that's left is gamepad stuff.  Additionally, the _getDeviceSlot function no longer exists.

PolygonalSun commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,308 @@
+     * @returns Current value of input
+     */
+    public getInput<T extends DeviceType>(type: T, inputIndex: DeviceInputs<T>, deviceSlot: number = this._firstDevice[type]): Nullable<number> {
+        if (!this._devices[type] || this._firstDevice[type] === undefined || this._devices[type][deviceSlot] === undefined) {
```
There are times when this will be undefined, such as when it's checking for a gamepad that isn't connected.

bghgary replied:
Does that means deviceSlot can be `undefined`?

PolygonalSun replied:
Yes.  If all devices for a given DeviceType are removed, deviceSlot could be undefined.

bghgary replied:
Okay, so deviceSlot currently has a type of number but it can also be undefined. Maybe you should default deviceSlot to undefined and set the deviceSlot manually after checking the _firstDevice.

PolygonalSun commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,308 @@
+     * @returns DeviceSource object
+     */
+    public getDeviceSource(deviceType: DeviceType, deviceSlot: number = this._firstDevice[deviceType]): Nullable<DeviceSource<DeviceType>> {
+        if (!this._devices[deviceType] || this._firstDevice[deviceType] === undefined || this._devices[deviceType][deviceSlot] === undefined) {
```
See previous comment related to this.

ryantrem commented on src/DeviceInput/deviceInputSystem.ts:
```diff
@@ -35,21 +13,21 @@ export class DeviceInputSystem implements IDisposable {
      * Callback to be triggered when a device is connected
      */
-    public onDeviceConnected: (deviceName: string) => void;
+    public onDeviceConnected: (deviceType: DeviceType, deviceSlot: number) => void = () => {};
```
I guess this means the native implementation will need to either duplicate the enum, or just use numbers directly, or consume the actual enum via napi. This seems like it complicates things a bit more for alternative implementations of `DeviceInputSystem`. Any thoughts on this @bghgary?

bghgary replied:
Meh, I doubt this enum is going to change much, so it's probably fine to duplicate it.

PolygonalSun replied:
If we're okay with the current implementation.  I'll just mark this as resolved.

PolygonalSun commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,308 @@
+    /**
+     * Default Constructor
+     * @param engine engine to pull input element from
+     * @param enableObserveEvents boolean to enable use of observe events
```
I didn't notice any performance issues when I was running this but I've been testing on a few reasonably powerful devices so I wasn't sure if my results were all that reliable for the average user.  In any case, I've removed the check and I'll revisit it later if need be.

bghgary replied:
We should at least capture a few profiles with these changes to make sure nothing is eating up too much resources.

PolygonalSun commented:
> Would it be possible to add an observable when input actually changed and have the user listen to this, instead of running it on the beforeRender loop?
> If I understand @ryantrem correctly, this is what he expected the API to look like (in the first comment) - more event driven.
> I know it is much simplified, but the WebXR motion controller API is also polling based. I added a few ways for the user to monitor changes, which you might be able to use here. And again, I understand this is a much broader API.
> 
> And just for me to understand - I noticed you removed observables and switched to callback-based API. Is there a specific reason for that?

To answer your questions, I've added a callback and observables to allow the user to receive and handle events that will fire off when input sources are updated.  The only caveat is that these will not be able to be used with Gamepads because they are not event based.  I may revisit this later on.  For the change back to callback from Observables, this was done to make the implementation of the Native version of the DeviceInputSystem easier.

bghgary commented on src/DeviceInput/InputDevices/deviceSourceManager.ts:
```diff
@@ -0,0 +1,186 @@
+     * @param deviceSlot "Slot" or index that device is referenced in
+     * @returns DeviceSource object
+     */
+    public getDeviceSource<T extends DeviceType>(deviceType: T, deviceSlot: number = this._firstDevice[deviceType]): Nullable<DeviceSource<T>> {
+        if (!this._devices[deviceType] || this._firstDevice[deviceType] === undefined || this._devices[deviceType][deviceSlot] === undefined) {
+            return null;
+        }
+
+        return this._devices[deviceType][deviceSlot];
+    }
```
```suggestion
    public getDeviceSource<T extends DeviceType>(deviceType: T, deviceSlot?: number): Nullable<DeviceSource<T>> {
        if (deviceSlot === undefined) {
            if (this._firstDevice[deviceType] === undefined) {
                return null;
            }

            deviceSlot = this._firstDevice[deviceType];
        }

        if (!this._devices[deviceType] || this._devices[deviceType][deviceSlot] === undefined) {
            return null;
        }

        return this._devices[deviceType][deviceSlot];
    }

```

PolygonalSun commented on src/DeviceInput/deviceInputSystem.ts:
```diff
@@ -67,28 +66,34 @@ export class DeviceInputSystem implements IDisposable {
      * Dispose of all the eventlisteners and clears the observables
      */
     public dispose() {
-        this.onDeviceConnectedObservable.clear();
-        this.onDeviceDisconnectedObservable.clear();
-
```
I converted those observables back to callbacks.  I'll update the comment.
